File: apps/android/app/src/main/java/com/variance/nearby/screens/profile/LeafNameInput.kt
package com.variance.nearby.screens.profile

/**
 * Pure validation of a Nearby leaf-name entry. A label is one or more `[a-z0-9-]` characters; a space
 * (or any other character) makes it invalid rather than being silently stripped. No network.
 */
sealed interface LeafNameInput {
    data object Empty : LeafNameInput

    /** A normalized (trimmed, lowercased) valid label. */
    data class Valid(val value: String) : LeafNameInput

    data object Invalid : LeafNameInput

    companion object {
        fun parse(raw: String): LeafNameInput {
            val trimmed = raw.trim()
            if (trimmed.isEmpty()) return Empty
            val lower = trimmed.lowercase()
            val valid = lower.all { it in 'a'..'z' || it in '0'..'9' || it == '-' }
            return if (valid) Valid(lower) else Invalid
        }
    }
}

File: apps/ios/nearby/Features/Profile/LeafNameInput.swift
import Foundation

/// Pure validation of a Nearby leaf-name entry. A label is one or more `[a-z0-9-]` characters; a space
/// (or any other character) makes it invalid rather than being silently stripped. No network.
enum LeafNameInput: Equatable {
  case empty
  /// A normalized (trimmed, lowercased) valid label.
  case valid(String)
  case invalid

  static func parse(_ raw: String) -> LeafNameInput {
    let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !trimmed.isEmpty else { return .empty }
    let lower = trimmed.lowercased()
    let valid = lower.allSatisfy { ($0.isLetter && $0.isASCII) || $0.isNumber || $0 == "-" }
    return valid ? .valid(lower) : .invalid
  }
}
